Constituent updates

peverwhee opened the pull request:
### Summary
Brings in optional metadata field to enable dynamic (run-time) constituents; also brings in a couple new methods for the constituent object.

closes #557 

### Description
Key dynamic constituent mods:
 - ccpp_capgen.py: create dynamic constituent dictionary (key=scheme name) and pass into API()
 - constituents.py: add handling of dynamic constituents to register constituents routine (including conditional call(s) to user-supplied routines); update error handling for register_constituents

Includes additional constituent object methods:
* set_water_species
* set_min_val
* set_molar_mass

Includes fixes to ensure consistent order of generated code (use statements, variables, etc)

Also brings back the .github/workflow/python.yaml workflow that got lost somehow

User interface changes?: Yes, but optional
User can now use "dynamic_constituent_routine" metadata header field to point to a function contained within the module fortran.

Testing:
  Modified advection test to include dynamic constituent routines.
  Added doctests to ccpp_datafile

### Generated code snippet
```
   subroutine test_host_ccpp_register_constituents(host_constituents, errflg, errmsg)
      ! Create constituent object for suites in 

      use ccpp_constituent_prop_mod, only: ccpp_constituent_properties_t
      ! Suite constituent interfaces
      use ccpp_cld_suite_cap,        only: cld_suite_constituents_num_consts
      use ccpp_cld_suite_cap,        only: cld_suite_constituents_const_name
      use ccpp_cld_suite_cap,        only: cld_suite_constituents_copy_const
      ! Dynamic constituent routines
      use cld_liq, only: cld_liq_dynamic_constituents
      use cld_ice, only: cld_ice_dynamic_constituents

      ! Dummy arguments
      type(ccpp_constituent_properties_t), target, intent(in)  :: host_constituents(:)
      character(len=512), intent(out)   :: errmsg ! ccpp_error_message
      integer,            intent(out)   :: errflg ! ccpp_error_code
      ! Local variables
      integer                                      :: num_suite_consts
      integer                                      :: num_consts
      integer                                      :: num_dyn_consts
      integer                                      :: index, index_start
      integer                                      :: field_ind
      type(ccpp_constituent_properties_t), pointer :: const_prop
      ! dynamic constituent props variable for cld_ice
      type(ccpp_constituent_properties_t), allocatable :: dyn_const_prop_0(:)
      ! dynamic constituent props variable for cld_liq
      type(ccpp_constituent_properties_t), allocatable :: dyn_const_prop_1(:)

      errflg = 0
      num_consts = size(host_constituents, 1)
      num_dyn_consts = 0
      ! Number of suite constants for cld_suite
      num_suite_consts = cld_suite_constituents_num_consts(errflg=errflg, errmsg=errmsg)
      if (errflg /= 0) then
         return
      end if
      num_consts = num_consts + num_suite_consts
      ! Add in dynamic constituents
      call cld_ice_dynamic_constituents(dyn_const_prop_0, errcode=errflg, errmsg=errmsg)
      if (errflg /= 0) then
         return
      end if
      num_dyn_consts = num_dyn_consts + size(dyn_const_prop_0)
      call cld_liq_dynamic_constituents(dyn_const_prop_1, errcode=errflg, errmsg=errmsg)
      if (errflg /= 0) then
         return
      end if
      num_dyn_consts = num_dyn_consts + size(dyn_const_prop_1)
      num_consts = num_consts + num_dyn_consts
      ! Pack dynamic_constituents array
      allocate(dynamic_constituents(num_dyn_consts), stat=errflg)
      if (errflg /= 0) then
         errmsg = 'failed to allocate dynamic_constituents'
         return
      end if
      index_start = 0
      do index = 1, size(dyn_const_prop_0, 1)
         dynamic_constituents(index + index_start) = dyn_const_prop_0(index)
      end do
      index_start = size(dyn_const_prop_0, 1)
      deallocate(dyn_const_prop_0)
      do index = 1, size(dyn_const_prop_1, 1)
         dynamic_constituents(index + index_start) = dyn_const_prop_1(index)
      end do
      index_start = size(dyn_const_prop_1, 1)
      deallocate(dyn_const_prop_1)
      ! Initialize constituent data and field object
      call test_host_constituents_obj%initialize_table(num_consts)
      ! Add host model constituent metadata
      do index = 1, size(host_constituents, 1)
         const_prop => host_constituents(index)
         call test_host_constituents_obj%new_field(const_prop, errcode=errflg, errmsg=errmsg)
         nullify(const_prop)
         if (errflg /= 0) then
            return
         end if
      end do

      ! Add dynamic constituent properties
      do index = 1, size(dynamic_constituents, 1)
         const_prop => dynamic_constituents(index)
         call test_host_constituents_obj%new_field(const_prop, errcode=errflg, errmsg=errmsg)
         nullify(const_prop)
         if (errflg /= 0) then
            return
         end if
      end do
      ! Add cld_suite constituent metadata
      num_suite_consts = cld_suite_constituents_num_consts(errflg=errflg, errmsg=errmsg)
      if (errflg /= 0) then
         return
      end if
      do index = 1, num_suite_consts
         allocate(const_prop, stat=errflg)
         if (errflg /= 0) then
            errmsg = "ERROR allocating const_prop"
            return
         end if
         call cld_suite_constituents_copy_const(index, const_prop, errflg=errflg, errmsg=errmsg)
         if (errflg /= 0) then
            return
         end if
         call test_host_constituents_obj%new_field(const_prop, errcode=errflg, errmsg=errmsg)
         nullify(const_prop)
         if (errflg /= 0) then
            return
         end if
      end do

      call test_host_constituents_obj%lock_table(errcode=errflg, errmsg=errmsg)
      if (errflg /= 0) then
         return
      end if
! Set the index for each active constituent
      do index = 1, SIZE(test_host_model_const_indices)
         call test_host_constituents_obj%const_index(field_ind,                                   &
              test_host_model_const_stdnames(index), errcode=errflg, errmsg=errmsg)
         if (errflg /= 0) then
            return
         end if
         if (field_ind > 0) then
            test_host_model_const_indices(index) = field_ind
         else
            errflg = 1
            errmsg = 'No field index for '//trim(test_host_model_const_stdnames(index))
            return
         end if
      end do

```

Commit:
Refactor constituent mixing ratio properties
Allow schema file to be passed to validate
Add Fortran comment write method
Add metadata table type and name to CCPP datatable variable entries
Improve generated constituent-handling code
Fix issues with output of long comments
Ensure more instance variables are 'private'
Have capgen create database object
Add line fill (target line length) interface to FortranWriter
New Fortran interface, indent_size
New Fortran interface, blank_line
Added insert function for verbatim copy to file
Allow arbitrary break for long lines
Make sure call list variables have an intent
Added link to constituent dictionary
Add metadata to CCPP constituent object DDT
Add Fortran writing unit tests and fix line break bugs
Added constituent props array host interface, improved unit conversion error
Use +ELLIPSIS instead of +IGNORE_EXCEPTION_DETAILS for doctests
Constituent cleanup, no function side effects in interface
Completed and optimized constituent accessor routines
Added minimum allowed value property

Commit:
Avoid RE deprecation warning by moving case fold to state_machine.py

Commit:
Fix doctest

Commit:
Allow '1' to be a valid unit

Commit:
Added molar_mass (molecular weight) value to constituent property type

Commit:
initial changes

Commit:
fixes

Commit:
address review comments

Commit:
hopefully final fixes

gold2718 committed:
Merge pull request #41 from peverwhee/add_molecular_weight

Add molecular weight as variable and constituent property
Fix missing property (advected) from database

Commit:
initial commit

Commit:
fix function name

Commit:
fix fstring

Commit:
fix mistakes in routine names and comments

Commit:
split up register and initialize

Commit:
cleanup

Commit:
cleanup

Commit:
add temporary hash_table variable to avoid dangling pointer

Commit:
deallocate hash_table

Commit:
cleaner way to deallocate hash table in iterator object

Commit:
change name of const array interfaces

Commit:
add intialized_in_physics property

Commit:
better error handling; fix for copy constituent

Commit:
remove unnecessary use statment

Commit:
fix unit tests

Commit:
revert changes for unit conversion test

Commit:
handle constituent edge cases better and differently; revert module variable in hash_table

Commit:
fix decimal

nusbaume committed:
Add 'thermo_active' property to constituent properties DDT.

nusbaume committed:
Update CCPP constituent standard names.

Commit:
Merged changes from new_variables_input
Rename dir for var_action test to distinguish from ct_build

Commit:
remove initialized-in-physics constituent from advection test required variable arrays

nusbaume committed:
Fix thermo_active syntax errors caught by unit tests.

nusbaume committed:
Add 'thermo_active' property tests to 'advection_test' unit test collection.

Commit:
Initial commit of containerized GitHub actions to run unit tests.

mwaxmonsky committed:
Fixing typo from previous commit.

mwaxmonsky committed:
Fixing typo from previous commit.

mwaxmonsky committed:
Addming missing python dependency.

mwaxmonsky committed:
Removing incorrect flag from pip install.

mwaxmonsky committed:
Replacing incorrect Dockerfile instruction.

mwaxmonsky committed:
Returning non-zero if test failures detected.

mwaxmonsky committed:
Removing test failing due to not being fully implemented yet.

mwaxmonsky committed:
Addming missing newlines.

Commit:
add default_value to constituent properties object

mwaxmonsky committed:
Adding testing CI action to pull requests and all branches.

mwaxmonsky committed:
Updating workflow name with more accurate description.

mwaxmonsky committed:
Renaming capgen unit test workflow file.

mwaxmonsky committed:
Fixing typo.

mwaxmonsky committed:
Moving capgen Dockerfile into test directory.

mwaxmonsky committed:
Adding print statement to test output for reasong for skipping var_action_test.

mwaxmonsky committed:
Moving tests to running in github container directly.

mwaxmonsky committed:
Replacing apt calls with apt-get.

mwaxmonsky committed:
Adding sudo to apt-get commands.

mwaxmonsky committed:
Updating package names to ubuntu packages.

mwaxmonsky committed:
Updating path to cd into.

mwaxmonsky committed:
Removing test/Dockerfile as no longer needed.

mwaxmonsky committed:
Removing un-needed python dependencies install step.

mwaxmonsky committed:
Removing sourcing python venv.

mwaxmonsky committed:
Filtering CI action to only auto run if in a PR or pushing to the main repo.

mwaxmonsky committed:
Testing allowing run of actions manually.

mwaxmonsky committed:
Adding proper filter to job to run unit test action manually.

mwaxmonsky committed:
Fixing format of workflow_dispatch declaration.

mwaxmonsky committed:
Changing ordering of workflow triggers.

mwaxmonsky committed:
Adding initial pylint configuration.

mwaxmonsky committed:
Removing pylint job to evaluate in a new PR.

mwaxmonsky committed:
Merge pull request #2 from mwaxmonsky/test_actions

Unit test actions
Adds capgen unit tests to the list of actions.

Commit:
add has_default method

Commit:
Merge branch 'add_const_interface' of https://github.com/peverwhee/ccpp-framework into add_const_interface

mwaxmonsky committed:
Attempt to add more python versions to python tests.

mwaxmonsky committed:
Updating action versions to no longer use deprecated node12.

mwaxmonsky committed:
Updating python.yaml to run on demand.

mwaxmonsky committed:
Temporarily adding push events back.

mwaxmonsky committed:
Updating metadata tests to check updated property name.

mwaxmonsky committed:
Initial doctest action and associated fixes.

mwaxmonsky committed:
Fixing actions file syntax error.

mwaxmonsky committed:
Fixing doctest call.

mwaxmonsky committed:
Fixing shell script syntax error.

mwaxmonsky committed:
Removing doctest from main unit test script as no longer needed with new doctest action.

mwaxmonsky committed:
Fixing no newline in some files.

mwaxmonsky committed:
Removing push events.

mwaxmonsky committed:
Initial attempt of removing python 2 code checks.

mwaxmonsky committed:
Removing python code older than 2.7

mwaxmonsky committed:
Chaning python unit tests to run in pytest harness.

mwaxmonsky committed:
Removing un-needed pylint statements and fixing imports.

mwaxmonsky committed:
Fixing formatting.

mwaxmonsky committed:
Removing python2 style is string check.

mwaxmonsky committed:
Update minimum version to 3.7

mwaxmonsky committed:
Merge pull request #4 from mwaxmonsky/remove_python_2

Initial removing python 2 and < 3.7 code checks.

mwaxmonsky committed:
Update .github/workflows/python.yaml

Updating github actions to v4.

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

mwaxmonsky committed:
Update .github/workflows/python.yaml

Update setup-python github action to v4.

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

mwaxmonsky committed:
Update .github/workflows/python.yaml

Removing un-needed whitespace.

Co-authored-by: Jesse Nusbaumer <nusbaume@ucar.edu>

mwaxmonsky committed:
Removing un-needed main function that runs doc test.

mwaxmonsky committed:
Removing un-needed main function that runs doctest.

mwaxmonsky committed:
Adding more python versions to test against.

mwaxmonsky committed:
Testing version syntax fix.

mwaxmonsky committed:
Testing updated version syntax fix.

mwaxmonsky committed:
Fixing version syntax.

mwaxmonsky committed:
Reverting version support syntax.

mwaxmonsky committed:
Removing pytest from testing of generated fortran script and removing doctest script as no longer usable.

mwaxmonsky committed:
Updating pytest.ini to ignore non-unit test files.

mwaxmonsky committed:
Adding verbose pytest output.

mwaxmonsky committed:
Re-adding unit test main check to unit tests.

mwaxmonsky committed:
Renaming test script to more appropriate name.

mwaxmonsky committed:
removing un-needed python requirements.txt install step.

mwaxmonsky committed:
Updating test script name missing from previous commit.

mwaxmonsky committed:
Removing requirements.txt as not needed currently.

peverwhee committed:
Merge pull request #1 from gold2718/add_const_interface

Merged changes from new_variables_input

mwaxmonsky committed:
Fixing typo

Commit:
fix advection test

mwaxmonsky committed:
Merge branch 'add_const_interface' into pytest_additions

mwaxmonsky committed:
Typo fix.

mwaxmonsky committed:
Merge pull request #3 from mwaxmonsky/pytest_additions

Updates python doctest implementations to use pytest as the harness and now uses pytest to run python unit tests in GitHub action.

Commit:
add new host cap interface to check if a variable is already a scheme constituent

Commit:
Merge branch 'add_const_interface' of https://github.com/peverwhee/ccpp-framework into add_const_interface

Commit:
add testing

nusbaume committed:
Make capgen unit tests less inter-dependent, and perform some cleanup.

nusbaume committed:
Merge remote-tracking branch 'peverwhee/add_const_interface' into const_updates

nusbaume committed:
Fix variable intent.

nusbaume committed:
Merge pull request #5 from nusbaume/const_updates

Add "thermodynamically active" property

Commit:
resolve merge conflicts

Commit:
fix check_errflg call

Commit:
address reviewer comments

Commit:
fix function calls

peverwhee committed:
Merge pull request #6 from peverwhee/new-interface

New "is_scheme_constituent" interface

nusbaume committed:
Set values for error flag and error message in new is_scheme_const subroutine.

peverwhee committed:
Merge pull request #7 from nusbaume/errflg_fix

Add missing error flag and error message

nusbaume committed:
Add new 'set_minimum' method which can change min constituent value

nusbaume committed:
Add unit tests for new set_minimum method.

nusbaume committed:
Fixed typos in test_host messages.

nusbaume committed:
Merge pull request #8 from nusbaume/qmin_setter

Add minimum value setter method.

nusbaume committed:
Add new 'water_species' property and associated methods.

nusbaume committed:
Add 'water_species' property unit tests.

nusbaume committed:
Fix comments in water species routines.

nusbaume committed:
Add space between '.not.' and logical to improve readability.

nusbaume committed:
Merge pull request #9 from nusbaume/water_species_prop

Add new water species constituent property

Commit:
sort sets before writing to files to ease comparison between runs

Commit:
fix attempt to sort list

nusbaume committed:
Merge pull request #10 from peverwhee/sorted_sets

Sort sets before writing

nusbaume committed:
Add setter methods for molecular weight, and add some additional property checks.

nusbaume committed:
Add new unit tests, and fix bugs found during testing.

nusbaume committed:
Replace 'molec_weight' with 'molar_mass', as it is likely a more accurate term.

nusbaume committed:
Merge pull request #12 from nusbaume/mol_weight_set

Add molecular weight property functionality

Commit:
initial commit; metadata table and parsing working

Commit:
code works; may need cleanup

Commit:
fix comment order - CRUCIAL

Commit:
merge to head of feature/capgen

Commit:
Merge branch 'dynamic_constituents' into peverwhee-fork-dynamic-constituents

Commit:
clean-up and switch constituents.py to fstrings

Commit:
expand testing

Commit:
update register_constituents error handling; misc review requests

Commit:
add doctests to ccpp_datafile.py

Commit:
review comments; add in vertical transform bugfix

peverwhee committed:
Merge pull request #13 from peverwhee/peverwhee-fork-dynamic-constituents

Add optional metadata field for dynamic constituent routine

Commit:
merge to head of feature/capgen

Commit:
fix merge

Commit:
another merge fix

Commit:
merge to head of feature/capgen

Commit:
Merge remote-tracking branch 'NCAR/main' into add_const_interface

Commit:
merge in NCAR/main

Commit:
fix merge

Commit:
fix test; bring back python.yaml workflow

dustinswales commented:
@peverwhee dyn_const_prop_0 and dyn_const_prop_1 are defined with the target attribute, but it's not needed. Is this intentional, or possibly an oversight/err by me in my previous commit?

Commit:
remove target attribute from dyn_const_prop variable declarations

peverwhee commented:
@dustinswales not intentional! I removed the "target" attribute

gold2718 commented on scripts/constituents.py:
```diff
@@ -484,124 +468,171 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+            # end for
+            cap.write("num_consts = num_consts + num_dyn_consts", 2)
+            cap.comment("Pack dynamic_constituents array", 2)
+            cap.write(f"allocate(dynamic_constituents(num_dyn_consts), stat={herrcode})", 2)
```
I do not understand the need for `dynamic_constituents`. Why not just copy out of the arrays returned by the schemes?

peverwhee replied:
I remember now (after trying to remove it) why I passed in the dynamic_constituents array from the host model. It's because the dyn_const_prop_x arrays go out of scope when we leave the register_constituents routine. Having the constituent properties array point to dynamic constituents that are declared at the host level keeps them in scope.

That said, there certainly could be a better way to do all of this!

gold2718 replied:
It seems inconvenient to ask the host model to do memory management for the framework. There could be a module variable for each suite's dynamic constituents.

Speaking of multiple suites, does this routine work if more than one suite is built in? For the solution above, I would expect one (allocateable or pointer) module variable per suite.

peverwhee replied:
Updated to use a single dynamic constituent array at the host cap level because it seemed easier.

To answer your other question: it does work if there's more than one suite - there are just more dynamic constituent routines called within register.

peverwhee replied:
Now have this at the top of <host>_ccpp_cap.F90:

`type(ccpp_constituent_properties_t), allocatable, target :: test_host_dynamic_constituents(:)`

And we're no longer passing in dynamic_constituents from the host and instead allocating/packing the test_host_dynamic_constituents array.

gold2718 replied:
I guess my concern is for a model such as UFS, there will be two suites but they do not have the same data. I don't know if they currently have non-advected constitutents but imagine that the physics has some of those. The "fast physics" suite that runs in the dycore only has the advected constituents and operates on different data so I think we will need multiple dynamic constituents arrays.

On the other hand, models such as the UFS will probably need to present as two host models which would lead to two host caps. Given the different "host data", I think this makes the most sense but we need to figure out how we are going to package and process the combined suites. Someone from NOAA correct me if I'm off base here.

In any case, we need to decide what types of arrangements will be supported to see if your current design will work in all those cases.

peverwhee replied:
I _think_ I understand, and based on that limited understanding, I would think that, yes, models like UFS will need two host caps (we'll need two constituent objects, not just two dynamic constituent arrays).

peverwhee replied:
opened related issue #561

gold2718 replied:
> I _think_ I understand, and based on that limited understanding, I would think that, yes, models like UFS will need two host caps (we'll need two constituent objects, not just two dynamic constituent arrays).

Maybe open an issue on this? This is sort of a combo framework <==> host model issue so not sure where it goes.

gold2718 reviewed (changes requested):
This is a partial review of a4548a0ebcf2f84d6940733f4021743b440588f0
I have some questions.

mkavulich commented on .github/workflows/python.yaml:
```diff
@@ -0,0 +1,49 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11']
```
Other CI tests go up to Python 3.12, is there a reason we can't also do that here?

peverwhee replied:
Good suggestion. I added 3.12 for the build tests, but ran into an annoying issue for the doctests and backed out of that - issue here: https://github.com/NCAR/ccpp-framework/issues/557

peverwhee replied:
Per our discussion yesterday, I changed the tests to no longer rely on the string representation of OrderedDict so now 3.12 is back in business!

Commit:
code cleanup; review requests; include python3.12 in ci testing

Commit:
merge to head of main

Commit:
back off 3.12 doctests for now

Commit:
add script to check fortran vs metadata without a host model

Commit:
add check that routine exists in fortran; use module-level dynamic constituents variable in cap rather than provided by host

Commit:
fix typo

Commit:
code cleanup; code review

Commit:
allow for skipping of ddt check; skip ddt check from offline script

mwaxmonsky reviewed (changes requested):
Looks great @peverwhee!

Has a few suggestions and some things I just wanted to clarify but just let me know if these are not relevant and I can re-review.  Thanks!

Commit:
handle dimensions for promoted variables

gold2718 commented on scripts/host_cap.py:
```diff
@@ -204,6 +227,37 @@ def constituent_model_const_index(host_model):
+###############################################################################
+    """Return the name of the function that will return a pointer to the
+       array of all constituents"""
+    hstr = "{}_constituents_array".format(host_model.name)
```
This suggestion (line 235) needs an initial `f` to make it an f-string.

mwaxmonsky replied:
Thanks @gold2718!  Updated my comment.

peverwhee replied:
converted to fstring!

gold2718 commented on scripts/host_cap.py:
```diff
@@ -204,6 +227,37 @@ def constituent_model_const_index(host_model):
+###############################################################################
+    """Return the name of the function that will return a pointer to the
+       array of advected constituents"""
+    hstr = "{}_advected_constituents_array".format(host_model.name)
```
This suggestion (line 243) needs an initial `f` to make it an f-string.

peverwhee replied:
converted to fstring

gold2718 commented on scripts/host_cap.py:
```diff
@@ -204,6 +227,37 @@ def constituent_model_const_index(host_model):
+def constituent_model_const_props(host_model):
+###############################################################################
+    """Return the name of the array of constituent property object pointers"""
+    hstr = "{}_model_const_properties".format(host_model.name)
```
This suggestion (line 250) needs an initial `f` to make it an f-string.

peverwhee replied:
converted to fstring

gold2718 commented on scripts/host_cap.py:
```diff
@@ -204,6 +227,37 @@ def constituent_model_const_index(host_model):
+###############################################################################
+    """Return the name of the interface that returns the array index of
+       a constituent array given its standard name"""
+    hstr = "{}_const_get_index".format(host_model.name)
```
This suggestion (line 258) needs an initial `f` to make it an f-string.

peverwhee replied:
converted to fstring

Commit:
check for compatibility before adding constituent; add more testing

Commit:
add unit comparison

peverwhee commented:
> Some test suggestion:
> 
> * Test for handling of duplicate compatible constituent
> * Test for handling of duplicate  semi-compatible constituent (e.g., unit difference), that the framework cannot handle.
> * Test for handling of duplicate incompatible constituent

Added logic to the constituents object to check for compatibility before adding a new constituent. Right now, that logic only checks a logical and character properties (advected, water species, thermo active, units). And added the following tests to the advection_test:
* Compatible constituent
* incompatible constituent

Also added unit tests to test parsing and fortran vs metadata comparisons

Commit:
code review

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -526,7 +538,7 @@ def parse_scheme_files(scheme_filenames, run_env):
         mtables = parse_metadata_file(filename, known_ddts, run_env)
         fort_file = find_associated_fortran_file(filename)
-        ftables = parse_fortran_file(fort_file, run_env)
+        ftables, additional_routines = parse_fortran_file(fort_file, run_env)
```
A good suggestion! I hadn't thought about it but have now looked into it and it would be a bit of a lift to add to ftables due to the way parsing works...

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -536,8 +548,16 @@ def parse_scheme_files(scheme_filenames, run_env):
         for sect in [x.sections() for x in ftables]:
             fheaders.extend(sect)
         # end for
+        dyn_routines = []
+        for table in mtables:
+            if table.dyn_const_routine:
+                dyn_routines.append(table.dyn_const_routine)
+            # end if
+        # end for
```
Great question - I think there's a possibility of multiple modules within a single metadata file and therefore multiple dynamic constituent routines.

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +649,31 @@ def capgen(run_env, return_db=False):
     # end if
     # Next, parse the scheme files
+    # We always need to parse the ccpp_constituent_prop_ptr_t DDT
+    ##XXgoldyXX: Should this be in framework_env.py?
```
comment removed

peverwhee commented on scripts/ccpp_datafile.py:
```diff
@@ -288,11 +400,73 @@ def _retrieve_dependencies(table):
+    ccpp_datafile.CCPPDatatableError: Could not find 'dyn_const_routines' element
+
+    """
+    result = set()
```
yep! removed.

peverwhee commented on scripts/host_cap.py:
```diff
@@ -101,6 +101,14 @@ def constituent_initialize_subname(host_model):
+    """Return the name of the subroutine used to initialize the
+    constituents for this run.
+    Because this is a user interface API function, the name is fixed."""
+    return "{}_ccpp_initialize_constituents".format(host_model.name)
```
converted to fstring!

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +629,29 @@ def capgen(run_env, return_db=False):
+               dyn_const_dict[table] = routine_name
+            else:
+               # dynamic constituent routines must have unique names
+               scheme_name = list(dyn_const_dict.keys())[list(dyn_const_dict.values()).index(routine_name)]
```
dyn_val_dict added.

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +629,29 @@ def capgen(run_env, return_db=False):
+    # end if
     scheme_headers, scheme_tdict = parse_scheme_files(scheme_files, run_env)
+    # Pull out the dynamic constituent routines, if any
+    dyn_const_dict = {}
```
dyn_val_dict added.

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +629,29 @@ def capgen(run_env, return_db=False):
+        routine_name = scheme_tdict[table].dyn_const_routine
+        if routine_name is not None:
+            if routine_name not in dyn_const_dict.values():
+               dyn_const_dict[table] = routine_name
```
dyn_val_dict added.

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +629,29 @@ def capgen(run_env, return_db=False):
+    for table in scheme_tdict:
+        routine_name = scheme_tdict[table].dyn_const_routine
+        if routine_name is not None:
+            if routine_name not in dyn_const_dict.values():
```
dyn_val_dict added

Commit:
update logic to exclude nested subroutines from routine parsing

Commit:
update vardictionary doctests and add 3.12 testing

climbfuji commented on test/advection_test/test_host.F90:
```diff
@@ -462,6 +576,154 @@ subroutine test_host(retval, test_suites)
+
+      !-------------------
+
+      !-------------------
```
Have we decided on a style for comments? I personally prefer a whitespace after the `!`. The `black` formatter used for Python code in spack also mandates white spaces after `#`.

peverwhee replied:
I agree that the white space after the "!" looks best. Updated comments in this file

climbfuji commented on test/unit_tests/sample_scheme_files/duplicate_dyn_const.F90:
```diff
@@ -0,0 +1,96 @@
+! Test parameterization with no vertical level
```
Here is a comment with a whitespace

climbfuji reviewed (changes requested):
One comment I have is that this PR is quite big and it contains a lot of somewhat unrelated changes. It would be better imo to separate out these different changes into individual PRs: f-strings, sorted, the changes to the water/minval/units, and the dynamic constituents?

I am not suggesting to break up this PR, but maybe consider submitting smaller, less intimidating to review changes in the future (in full awareness that I am guilty of these catchall PRs from time to time ...).

I am also not suggesting to make any changes to the style of the comments. This could be a separate PR where we clean up just the formatting and re-enable the linter(s) in CI.

Overall this looks very good to me, but I don't understand it well enough. I requested a couple of small changes and I will approve afterwards, but someone with a better understanding should be reviewing and approving this (and of course the full testing needs to be done once the reviews are finalized) before it gets merged.

Thanks for all the hard work that went into this PR!

Commit:
Merge remote-tracking branch 'NCAR/main' into group-fix

Commit:
code cleanup; review requests

Commit:
code clean-up; add test for promoted variable with no horiz dimension

Commit:
code review; cleanup

Commit:
Merge remote-tracking branch 'NCAR/main' into add_const_interface

Commit:
only run pytests on test/ directory in python.yaml workflow

peverwhee commented on pytest.ini:
```diff
@@ -0,0 +1,2 @@
+[pytest]
+addopts = -ra --ignore=scripts/metadata2html.py --ignore-glob=test/**/test_reports.py
```
added!

peverwhee commented on scripts/ccpp_datafile.py:
```diff
@@ -383,7 +599,48 @@ def _retrieve_variable_list(table, suite_name,
     If <intent_type> is present, return only that variable type (input or
     output).
-    If <excl_prot> is True, do not include protected variables"""
+    If <excl_prot> is True, do not include protected variables
```
I know you explicitly said not in this PR, but i went ahead and made the change - find/replace eemed easier than making an issue (but i see how things like this make the PR more unwieldy...)

peverwhee commented on src/ccpp_hash_table.F90:
```diff
@@ -396,7 +396,7 @@ subroutine hash_iterator_initialize(this, hash_table)
       ! Dummy arguments
       class(ccpp_hash_iterator_t)      :: this
-      class(ccpp_hash_table_t), target :: hash_table
+      type(ccpp_hash_table_t), intent(in), target :: hash_table
```
Good question! Must've been a relic of some debugging I was doing  - changed back

climbfuji reviewed (approved):
Thanks for addressing my comments. This is a non-expert approval which on its own is not sufficient for merging ;-)

Commit:
merge in optional fortran v metadata checker

Commit:
Merge remote-tracking branch 'origin/group-fix' into add_const_interface

Commit:
Merge remote-tracking branch 'NCAR/main' into add_const_interface

dustinswales commented on src/ccpp_constituent_prop_mod.F90:
```diff
@@ -989,13 +1161,35 @@ subroutine ccp_model_const_add_metadata(this, field_data, errcode, errmsg)
-!!XXgoldyXX: Add check on key to see if incompatible item already there.
+         ! Check to see if standard name is already in the table
+         call field_data%standard_name(standard_name, errcode, errmsg)
+         cprop => this%find_const(standard_name)
```
I see that you nullify cprop at the start of this%find_const(). 
Is it best to pass in a null pointer, and return the null pointer if condition not met, otherwise return associated pointer. A purely stylistic question.

peverwhee replied:
An interesting point. I would have to modify cprop to be inout and change all the calls to do this, so I'm leaning towards not changing it.

Willing to be overruled, though.

dustinswales replied:
Good point. It's fine as is.

dustinswales commented on src/ccpp_constituent_prop_mod.F90:
```diff
@@ -989,13 +1161,35 @@ subroutine ccp_model_const_add_metadata(this, field_data, errcode, errmsg)
+         cprop => this%find_const(standard_name)
+         if (associated(cprop)) then
+            ! Standard name already in table, let's see if the existing constituent is the same
+            cindex = cprop%const_index()
```
Not used.

peverwhee replied:
removed!

dustinswales commented on test/advection_test/test_host.F90:
```diff
@@ -314,19 +318,63 @@ subroutine test_host(retval, test_suites)
+      call test_host_ccpp_deallocate_dynamic_constituents()
+      deallocate(host_constituents)
+      allocate(host_constituents(2))
+      call host_constituents(1)%instantiate(std_name="specific_humidity",     &
```
I probably don't understand the test here, but how is it allowed to have two host constituents with the same standard_name?

peverwhee replied:
It's allowed if they're identical (in which case it's just not added at all). This handles the case where two different schemes dynamically add the same constituent.

dustinswales commented on test/advection_test/test_host.F90:
```diff
@@ -357,212 +405,480 @@ subroutine test_host(retval, test_suites)
+         errflg = 0
+      end if
+      ! Check moist mixing ratio for a dynamic constituent
+      call const_props(index_dyn2)%is_dry(const_log, errflg, errmsg)
```
Should this be calling "is_moist"?

peverwhee replied:
added another testing code block to confirm that is_moist returns true (but keeping the test of is_dry() returning false)

dustinswales commented on test/unit_tests/sample_scheme_files/duplicate_dyn_const.F90:
```diff
@@ -0,0 +1,96 @@
+! Test parameterization with no vertical level
```
Can you update the description?
Also, is this test exercising duplicates?

peverwhee replied:
good point - updated the description. It's exercising duplicate dynamic constituent routine _names_

dustinswales commented on test/unit_tests/sample_scheme_files/dyn_const_not_present.F90:
```diff
@@ -0,0 +1,75 @@
+! Test parameterization with no vertical level
```
Can you update the description?

peverwhee replied:
nope! changed back

dustinswales commented on test/var_compatibility_test/run_test:
```diff
@@ -6,7 +6,7 @@ scriptdir="$( cd $( dirname $0 ); pwd -P )"
 ## Option default values
 ##
-defdir="ct_build"
+defdir="va_build"
```
Is this needed?

peverwhee replied:
nope! changed back.

dustinswales reviewed (changes requested):
@peverwhee Some minor comments, but overall looks good. Thanks for these changes!
I will start testing in the UFS using this branch and push to get on the commit queue asap.

Commit:
address review comments

peverwhee commented on src/ccpp_constituent_prop_mod.F90:
```diff
@@ -989,13 +1161,35 @@ subroutine ccp_model_const_add_metadata(this, field_data, errcode, errmsg)
-      character(len=*), parameter :: subname = 'ccp_model_const_add_metadata'
+      character(len=errmsg_len)                    :: error
+      character(len=*), parameter                  :: subname = 'ccp_model_const_add_metadata'
+      type(ccpp_constituent_properties_t), pointer :: cprop
```
done.

peverwhee commented on scripts/constituents.py:
```diff
@@ -484,124 +468,168 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+        cap.write(f"integer{spc} :: num_dyn_consts", 2)
+        cap.write(f"integer{spc} :: index, index_start", 2)
+        cap.write(f"integer{spc} :: field_ind", 2)
         cap.write(f"type({CONST_PROP_TYPE}), pointer :: const_prop", 2)
```
updated.

peverwhee commented on scripts/ccpp_capgen.py:
```diff
@@ -629,7 +679,24 @@ def capgen(run_env, return_db=False):
         scheme_files = [const_prop_mod] + scheme_files
     # end if
     # Next, parse the scheme files
+    # We always need to parse the ccpp_constituent_prop_ptr_t DDT
```
This just makes the ccpp constituents available to the host, but doesn't require the host to use it.

dustinswales replied:
Excellent. Thanks!

dustinswales reviewed (approved):
@peverwhee Thanks for addressing my comments. Looks good!
(Running the UFS tests now, and will open PRs)

grantfirl commented:
@gold2718 @mwaxmonsky Would you like to re-review before merging this?

mwaxmonsky reviewed (approved):
Apologies for the delay, looks good to me!

grantfirl commented:
> > @gold2718 @mwaxmonsky Would you like to re-review before merging this?
> 
> Sorry, can I have today?

Sure. We're trying to attach this to a set of UFS PRs that are on a queue to be final tested/merged early next week. If we can get final approval this week, it should be fine.

gold2718 reviewed (changes requested):
I have not looked at the test code yet but I have a couple of things that look like they should be changed (looking at 6e165e5c483ece98e3dfc407f1e4d67d1a5f5eaf).

peverwhee commented on scripts/constituents.py:
```diff
@@ -484,124 +468,168 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+           cap.write("end if", 3)
+           cap.write("end do", 2)
+        # end if
+        
```
we can't deallocate {dyn_const_name} here because that's the module-level variable that holds the data for the dynamic constituents; deallocating it would mean you could no longer access those constituent properties

Commit:
fix incorrect subname

peverwhee commented on src/ccpp_constituent_prop_mod.F90:
```diff
@@ -1865,6 +2101,28 @@ end subroutine ccpt_is_thermo_active
+      integer,          optional,         intent(out) :: errcode
+      character(len=*), optional,         intent(out) :: errmsg
+      ! Local variable
+      character(len=*), parameter :: subname = 'ccpt_is_thermo_active'
```
fixed!

gold2718 commented on scripts/constituents.py:
```diff
@@ -484,124 +468,168 @@ def write_host_routines(cap, host, reg_funcname, init_funcname, num_const_funcna
+                cap.write(f"do index = 1, size(dyn_const_prop_{idx}, 1)", 2)
+                cap.write(f"{dyn_const_name}(index + index_start) = dyn_const_prop_{idx}(index)", 3)
+                cap.write("end do", 2)
+                cap.write(f"index_start = size(dyn_const_prop_{idx}, 1)", 2)
```
I see how this works if there are two entries in `dyn_const_dict` but what about 3? Shouldn't this be:
```suggestion
                cap.write(f"index_start = index_start + size(dyn_const_prop_{idx}, 1)", 2)
```

peverwhee replied:
yes good point! fixed

Commit:
fix indexing

gold2718 reviewed (approved):
I'm still not sure I have covered all the corner cases (a simpler implementation would be easy to verify), but I have not found any more definite problems.
